SHIPP tool integration

rtesra opened the pull request:
To do:

- [x] Add tests
- [x] Add function to update excel workbook with outputs from `agyw_format_naomi()`
- [x] Add scale to KP consensus estimates from KP workbook
- [x] Add scale to new infections estimates from KP workbook or Goals ASM database
- [x] Test workbook on multiple countries

jeffeaton committed:
add function to read Spectrum KP table

rtesra committed:
initial commit

rtesra commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1475 @@
+
+
+
+agyw_generate_risk_populations <- function(naomi_output) {
```
@r-ash this functions takes in a naomi output zip or hintr object and retursn the three data.frames needed to update the agyw tool

r-ash committed:
Merge branch 'master' into mrc-4652

r-ash committed:
Add WIP

r-ash committed:
Add WIP impl

rtesra committed:
fix incidence calculations

rtesra committed:
remove survey data from prevalence calculation

rtesra committed:
update workflow to pull in updated KP estimates

r-ash committed:
Merge branch 'agyw-tool-integration' into mrc-4652

r-ash committed:
Use new format of workbook

rtesra committed:
add labels to aggregated data and patch for naomi.resource update

rtesra committed:
Merge branch 'agyw-tool-integration' into mrc-4652

rtesra committed:
Update test-downloads.R

rtesra committed:
update test data and fix bugs

rtesra committed:
Update to match renaming of columns in ASF data in naomi.resources

krisher1 committed:
Update agyw integration code

-Adds comments & sources
-Corrects minor issues w/translation of code
-Updates KP calculations based on updated data
-Corrects age distribution of MSM to reflect age at sexual debut by country

rtesra committed:
Merge branch 'master' into agyw-tool-integration

r-ash committed:
Wire up data with AGYW template

r-ash committed:
Fix spdep warning by explicitly setting style for mat2listw

r-ash commented on R/car.R:
```diff
@@ -74,7 +74,7 @@ scale_gmrf_precision <- function(Q,
                                  eps = sqrt(.Machine$double.eps)) {
 
-  nb <- spdep::mat2listw(abs(Q), style = "M")$neighbours
+  nb <- spdep::mat2listw(abs(Q), style = "B")$neighbours
```
Note I've updated this using Paddy's fix from https://github.com/mrc-ide/naomi/pull/391 which should stop the huge amount of warnings. We can close that PR after this has been merged

r-ash replied:
Actually tried that and it broke everything, just suppressing this warning on "M" for now as I can't find an equivalent to this which won't trigger a warning

r-ash committed:
Suppress single warning

rtesra committed:
Merge branch 'pr/419' into agyw-tool-integration

rtesra committed:
Merge branch 'agyw-tool-integration' of https://github.com/mrc-ide/naomi into agyw-tool-integration

rtesra committed:
align katie's code and clean up documentation

rtesra committed:
Merge remote-tracking branch 'origin/pjnz-kp-extract' into agyw-tool-integration

rtesra committed:
account for all label configurations

rtesra committed:
scale to spectrum kp workbook consensus estimates

rtesra committed:
remove tetsing files :(

rtesra committed:
Update NEWS.md

r-ash committed:
Use with_mocked_bindings over with_mock

r-ash committed:
Add note about with_mocked_bindings

r-ash committed:
Fix up test, regen docs

r-ash committed:
Set min naomi.resources version

osymandius commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1513 @@
+  if(!is.na(fsw_consensus)){
+
+    # Check if consensus estimate is larger than age matched population denominator
+    pop <- naomi_pop[naomi_pop$area_level == 0 & naomi_pop$age_group == "Y015_049" & naomi_pop$sex == "female",]$population
```
Would be happy to throw an error (or use the default admin1 PSE% from our analysis) in cases where consensus estimates are >10% of the sex/age matched denominator.

jeffeaton replied:
Will defer. 

What will happen if this fails though? I see in next line that it has a `stopifnot()`.  Should we manage the fail more gently?

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1517 @@
+}
+
+
+#' Calculate incidence in high risk male key populations
+#'
+#' @param outputs Naomi output.
+#' @param options Naomi options extracted from outputs
+#' @param male_srb Estimates of male sexual risk groups generated by `agyw_adjust_sexbehav_msm_pwid()`
+#' @param male_logit_prevalence Risk adjusted estimates of male prevalence in sexual risk groups generated by `agyw_calculate_prevalence_male()`
+#' @param survey_year NOTES:: add in description + when this should be adjusted. Hardcoded to 2018.
+#'
+#' @return Wide format output required for the AGYW workbook
+#'
+#' @export
+#'
+#' Survey year should be updated to most current household survey in the country -
+#' for countries without recent household surveys, leave at 2018 - the spatiotemporal
+#' model of sexual behaviour fitted to all countries has the most data for in roughly 2018
```
Can you fix the docs here, couple of things
* Update the parameters
* The details about "survey year" should either be under the parameter, after the title or under a `@details` section. This won't go into the generated docs where it is atm
* Does this need to be exported? I think it is only used internally in which case replace the `export` with `@keywords internal`

rtesra replied:
Thanks - have updated.

r-ash commented on tests/testthat/helper-testing.R:
```diff
@@ -3,6 +3,8 @@ expect_no_error <- function(object) {
 
 with_mock <- function(..., .parent = parent.frame()) {
+  ## Don't use this, this should be removed, now testthat have
+  ## added with_mocked_bindings that should be preferred
```
Thanks for adding this!

r-ash committed:
Move hintr agyw test helper into Naomi

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1517 @@
+#'
+#' @return Naomi indicators formatted for the AGYW workbook.
+#'
+#' @export
```
Does this actually need to be exported? Will you use it from anywhere other than Naomi?

rtesra replied:
Thanks - have removed.

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1517 @@
+#'
+#' @return District level FSW estimates by 5-year age bands for ages 15-49.
+#'
+#' @export
```
Same with this export?

rtesra replied:
Thanks - have removed.

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1517 @@
+                                           male_srb,
+                                           survey_year_sample = 2018) {
+
+  #' Naomi estimates of PLHIV and population by district and age band
```
```suggestion
  # Naomi estimates of PLHIV and population by district and age band
```

r-ash replied:
And elsewhere in this file

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1517 @@
+  #' Check that sum of disaggregated infections is the same as total infections
+  sum_infections <- df3$infections_nosex12m + df3$infections_sexcohab + df3$infections_sexnonreg + df3$infections_sexpaid12m
+
+  # TO DO: Flag this to add in warning - stop please contact support (usually an issue with mapping boundaries)
+  # ADD IN WARNIING HERE
+  stopifnot(max(df3$infections - sum_infections) < 10^{-9})
```
Do you want to do this as part of this PR? stopifnot's generally good to avoid as the error message they produce is not easy to understand by users

rtesra replied:
Have converted into more informative warning.

r-ash reviewed (changes requested):
This looks good, couple of notes about `#'` being used outside of function docs and use of `export`. I've not looked very deeply at the data manipulation stuff, agree that we should probably run this for a few countries and see if it highlights any edge cases we've not thought about.

r-ash commented on R/car.R:
```diff
@@ -74,7 +74,17 @@ scale_gmrf_precision <- function(Q,
+  # options for style seemingly do the same thing as style = "M"
+  # so just ignoring this warning for now
+  withCallingHandlers({
+    nb <- spdep::mat2listw(abs(Q), style = "M")$neighbours
```
There are new warnings coming from this.

In the current build

```
Warning (test-car.R:12:3): scale_gmrf_precision() returns expected values
6 is not an origin
Backtrace:
    ▆
 1. └─naomi::scale_gmrf_precision(Q) at test-car.R:12:3
 2.   ├─base::withCallingHandlers(...) at src/R/car.R:81:3
 3.   └─spdep::mat2listw(abs(Q), style = "M") at src/R/car.R:82:5
 4.     └─spdep::sn2listw(...)
Warning (test-car.R:12:3): scale_gmrf_precision() returns expected values
no-neighbour observations found, set zero.policy to TRUE;
this warning will soon become an error
Backtrace:
    ▆
 1. └─naomi::scale_gmrf_precision(Q) at test-car.R:12:3
 2.   ├─base::withCallingHandlers(...) at src/R/car.R:81:3
 3.   └─spdep::mat2listw(abs(Q), style = "M") at src/R/car.R:82:5
 4.     └─spdep::sn2listw(...)
```

Is there anything we can do about this? Can we replace this function entirely? It feels like this package is always chucking out warnings and there is a new one every few weeks.

rtesra replied:
@jeffeaton I thank we have discussed this in the past - any thoughts?

jeffeaton replied:
just confirming this is resolved now in #418.

rtesra committed:
Update R/agyw-integration.R

Co-authored-by: Rob <39248272+r-ash@users.noreply.github.com>

rtesra committed:
fix up documentation

rtesra committed:
Merge branch 'agyw-tool-integration' of https://github.com/mrc-ide/naomi into agyw-tool-integration

rtesra committed:
add checks for alignment between agyw resources and naomi estimates

r-ash committed:
Merge branch 'master' into agyw-tool-integration

r-ash committed:
Merge branch 'master' into agyw-tool-integration

r-ash committed:
Regen docs

r-ash committed:
Fix issue with incorrect function arg

r-ash committed:
Remove unused helper function

rtesra committed:
add agyw vignette

rtesra committed:
Squashed commit of the following:

commit 551ada7fa85e8753da155062d0faf12824dd4eb7
Author: Katie Risher <krisher@ic.ac.uk>
Date:   Fri Jan 19 16:53:00 2024 -0500

    Additional edits to the agyw integration functions

    -New infections instead of incidence for the Naomi outputs tab
    -Naming the countries to match the tool
    -some edits so the area names are formatted more nicely
    -max RR for FSW capped at 35 - if over 35 leads to impossible incidence estimates in some districts

commit 666c504621671089d8deadb32b93dc22989738ce
Author: Katie Risher <krisher@ic.ac.uk>
Date:   Sun Dec 17 22:08:52 2023 -0500

    Add outputs argument to agyw_calculate_prevalence_male

    Add outputs argument to agyw_calculate_prevalence_male

commit 0d3529618cea12c668335816512785f50a3f9078
Author: Katie Risher <krisher@ic.ac.uk>
Date:   Thu Dec 14 23:05:36 2023 -0500

    Remove rr_sexnonreg

    Remove rr_sexnonreg from female output

rtesra committed:
update code to clean up country names for workbook

rtesra commented on R/agyw-integration.R:
```diff
@@ -88,65 +88,64 @@ agyw_format_naomi <- function(outputs, options){
-
-
 
+  country_name_db <- tibble::tribble(~country, ~iso3,
```
@krisher1 I have updated the method for hardcoding the country names to pulling straight from a dataframe.

rtesra commented on R/agyw-integration.R:
```diff
@@ -192,9 +191,7 @@ agyw_format_naomi <- function(outputs, options){
-  if(options$area_scope=="COD") {
-    naomi_wide$Country <- "Democratic Republic of the Congo"
-  }
+
```
@krisher1 the change in code above should maintain the correct name for DRC.

jeffeaton commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+  }
+
+
+  # Assumption from literature that 9% of PWID are female so remove them from
+  # the male denominator
```
@osymandius: can we add any reference in the comment here to the 'literature' this is from?

rtesra replied:
@osymandius and @krisher1 following up on this?

jeffeaton commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+  pwid$total_pwid <- pwid$total_pwid * 0.91
+
+  # PWID age distribution
+  # Review of literature - Hines et al Lancet Global Health 2020
```
Thanks — add DOI or link if possible

rtesra replied:
Added here: https://github.com/mrc-ide/naomi/pull/413/commits/e923fceed2f0c943e3a3a6423fb6e9407bdd7d73

jeffeaton commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+
+  summarise_age_cat_female <- function(dat, age_cat) {
+
+    if (age_cat == "Y015_024") {age_groups <- c("Y015_019", "Y020_024")}
+    if (age_cat == "Y025_049") {age_groups <- c("Y025_029","Y030_034","Y035_039",
+                                                "Y040_044", "Y045_049")}
+    if (age_cat == "Y015_049") {age_groups <- c("Y015_019", "Y020_024","Y025_029",
+                                                "Y030_034","Y035_039","Y040_044",
+                                                "Y045_049")}
```
Similar comments to above; would prefer to avoid hardcoding these; fine to leave for now.

jeffeaton commented on R/downloads.R:
```diff
@@ -115,8 +115,18 @@ hintr_prepare_agyw_download <- function(output, pjnz,
-  dummy_data <- data.frame(x = c(1, 2, 3), y = c(3, 4, 5))
-  writexl::write_xlsx(list(sheet = dummy_data), path = path)
+
+  template_path <- naomi.resources::get_agyw_workbook_path()
+
+  risk_populations <- agyw_generate_risk_populations(output$model_output_path,
+                                                     pjnz)
+
+  sheets <- list(
+    "All outputs - F" = risk_populations$female_incidence,
+    "All outputs - M" = risk_populations$male_incidence,
+    "NAOMI outputs" = risk_populations$naomi_output
+  )
+  write_xlsx_sheets(template_path, sheets, path = path)
```
What happens if something fails / throws an error in here?

r-ash replied:
Thanks, I've written a wrapper which will add some more context should anything in excel writing error. Should return some more info to users https://github.com/mrc-ide/naomi/pull/413/commits/a10bfcd0a8f60407c69acf400e402f42736afb25#diff-8e11e9d56ce8b4fdd117a811bd64a4b2ea57306740475e9703d27faa1131cac3R1665

jeffeaton commented on man/agyw_adjust_sexbehav_fsw.Rd:
Do all of these functions need to be exported / have `man` files generated?  I see functions labelled as `@internal`.  I think if you add something like `@noRd` it will also suppress generating the `man` files (@r-ash can confirm).  Or maybe it's an intentional choice to generate them.

r-ash replied:
`@keywords internal` removes the function from the documentation index, `@noRd` means the roxygen docs aren't generated at all.

Looks like people to tend to use `@noRd` for internal functions, and `@keywords internal` for functions aimed at other developers, not typical users of the package.

So yep I agree, `@noRd` is better here.

jeffeaton reviewed (commented):
Thanks @rtesra and all. This is heroic.  Also clear and well commented, thanks.

I haven't checked details (obviously), but have reviewed enough that not concerned this will interact poorly with anything else.

I have left some styling comments in here. You can basically take / add to an issue for later / leave forever those at your discretion.

Two of the comments that I would like check: 

1. What will happen if `hintr_prepare_agyw_download()` fails? [What will user see?]
2. Are we intentionally generating `man` files for all the functions here?  I think can add `@noRd` to the Roxygen to suppress.

rtesra commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+      dplyr::group_by(area_id, area_name, area_level,spectrum_region_code,
+                      calendar_quarter, sex) %>%
+      dplyr::summarise(
+        "population" = sum(population * as.integer(age_group %in% age_groups)),
+        "plhiv" = sum(plhiv * as.integer(age_group %in% age_groups)),
+        "infections" = sum(infections * as.integer(age_group %in% age_groups)),
```
These aggregation do not exist in the indicators currently (15-29 and 25-49). The purpose of this function is formate and aggregate Naomi inputs for the rest of of the workflow - will add some additional context in in the functions description.

rtesra commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+
+  # Naomi indicators for 5-year age groups + 15-49
+  df2 <- naomi_ind_labelled %>%
+    dplyr::filter(age_group %in% c("Y015_019", "Y020_024", "Y025_029", "Y030_034",
+                                   "Y035_039", "Y040_044", "Y045_049", "Y015_049")) %>%
+    dplyr::select(names(df1)) %>%
```
Not following on how we would select these age groups using a join?

rtesra commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1638 @@
+                                     "Ethiopia", "ETH",
+                                     "Haiti", "HTI",
+                                     "Chad", "TCD",
+                                     "Cote D'Ivoire", "CIV",
```
@krisher1 I think that is for you - am I correct in thinking this is hardcoded to match the workbooks?

rtesra committed:
add DOI for Hines 2020 paper

rtesra committed:
add warning for bad KP workbook consensus estimates

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1659 @@
+    prop_pop <- fsw_consensus / pop
+
+    if(prop_pop >= 0.05) {
+      t_("FSW_CONSENSUS_WARNING")
```
This call doesn't go anywhere? Did you mean it as a waning

rtesra replied:
Have removed and fixed up test to check that this working.

r-ash commented on R/agyw-integration.R:
```diff
@@ -0,0 +1,1659 @@
+    prop_pop <- pwid_consensus / pop
+
+    if(prop_pop >= 0.05) {
+      t_("PWID_CONSENSUS_WARNING")
```
This also doesn't go anywhere

rtesra replied:
Have removed and fixed up test to check that this working.

r-ash commented:
> On warnings in general, Ashton, Robert we should add a clearer error for if the workbook fails to generate. So far we have the following warnings/errors in place:
Error: If area hierarchy if Naomi inputs does not match any of external data inputs. Flags which area_ids are inconsistent and tells user to contact support.
Warning: (I'll add this in now) If KP consensus estimate is larger than 5% of matched population denominator. Tells user that default proportions will be used instead. 
Error: If sum of disaggregated infections does not equal total infections - tells user to contact support. This shouldn't happen now that we flag the mismatched area_ids at the start but left it for edge cases (who knows).

Errors sounds good, I will add some specific error handling into the excel file generation to return a nicer message should this fail.

Note about the warnings, these are not surfaced at the moment for downloads. So fine to add these here if you want but users won't see them until we add some specific handling for this in the front end. It will require a new endpoint in hintr to get these surfaced.

In general also after seeing the other warnings I am reluctant to use them. I think people have been trained to ignore the warnings.

That said, see the note above about a few floating calls to `t_` which I think you want to wrap in a warning?

r-ash committed:
Bump required version of naomi.resource and update branch pin

r-ash committed:
Update write sheet to work with file which has existing column headers

r-ash committed:
Return error message with context should generation fail

rtesra committed:
fix failing test for scaling to kp workbook consensus estimate

rtesra committed:
remove warnings that are not output anywhere

r-ash committed:
Remove branch pin, set duckdb name in DESCRIPTION

rtesra committed:
Update agyw-integration.R

add placeholder for incidence adjustment code

krisher1 committed:
Add KP adjustment for new infections

Added KP adjustment to incidence estimates to match Spectrum KP new infection estimates for women & men - results in updated estimates of incidence for other behavioural groups as well

rtesra committed:
Merge branch 'agyw-tool-integration' into pse-tool-update-incidence

rtesra committed:
Merge pull request #430 from krisher1/pse-tool-update-incidence

Add KP adjustment for new infections

rtesra committed:
rename agyw -> shipp

rtesra committed:
Scale new infection to KP workbook or GOALS

krisher1 committed:
Update SHIPP vignette + some code fixes

- Align SHIPP vignette with the methods used
- Minor code fixes for adjusting KP new infections to either KP workbook or Goals estimates under the assumption that a consensus KP new infection estimate is never missing

rtesra committed:
Merge branch 'master' into pse-tool-update-incidence

krisher1 committed:
Sum infections_pwid not pwid

krisher1 committed:
Merge branch 'pse-tool-update-incidence' of https://github.com/mrc-ide/naomi into pse-tool-update-incidence

rtesra committed:
Merge branch 'master' into agyw-tool-integration

r-ash committed:
Force installation of latest naomi.resources from github

r-ash committed:
Bump version number

rtesra committed:
edge case for 0 pse

rtesra committed:
Merge branch 'agyw-tool-integration' of https://github.com/mrc-ide/naomi into agyw-tool-integration

rtesra committed:
aggregate KP workbook from subnational pjnz files

rtesra committed:
clean up docs

rtesra committed:
Merge branch 'master' into agyw-tool-integration

rtesra committed:
Revert "Merge branch 'master' into agyw-tool-integration"

This reverts commit d4a26c8fd686dbbf8ff3c7a8b6490e0cfe5f9ed8, reversing
changes made to 68e2b8ade60339b2e9907137a6a11c7b3c04e3ea.
